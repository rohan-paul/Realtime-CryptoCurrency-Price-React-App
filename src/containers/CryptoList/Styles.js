import React from "react"
import { makeStyles } from "@material-ui/core/styles"

const useStyles = makeStyles(theme => ({
  container: {
    margin: "auto",
    backgroundColor: theme.background.default,
  },
  table: {
    backgroundColor: theme.background.paper,
    borderRadius: theme.shape.borderRadius,
    paddingBottom: "100px",
    alignItems: "center ",
    justifyContent: "center ",
    verticalAlign: "middle ",
    textAlign: "center",
  },
  inputandButtonContainer: {
    display: "flex",
    flexDirection: "row",
    alignItems: "center ",
    justifyContent: "center ",
  },
  selectionBox: {
    marginTop: "20px",
    marginRight: "10px",
    width: "55%",
    height: "50px",
  },
  button: {
    marginTop: "20px",
    width: "25%",
    height: "50px",
  },

  spinner: {
    height: "calc(100vh - 100px)",
  },
}))

export { useStyles }
