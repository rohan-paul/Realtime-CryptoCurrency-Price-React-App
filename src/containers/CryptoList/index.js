/* eslint-disable jsx-a11y/no-static-element-interactions */
/* eslint-disable jsx-a11y/click-events-have-key-events */
/* eslint-disable react/button-has-type */
/* eslint-disable-next-line react-hooks/exhaustive-deps */
import React, { useEffect, useState, useRef } from "react"
import { useSelector, useDispatch } from "react-redux"
import PropTypes from "prop-types"
import { Typography } from "@material-ui/core"
import GlobalSnackbar from "../../components_libs/GlobalSnackbar"
import LoadingSpinner from "../../components_libs/LoadingSpinner"
import {
  loadCurrencyList,
  handleUserCurrencySelection,
  handleSnackBarStatus,
  getSelectedCurrency,
} from "../../actions/getUserActions"
import Button from "@material-ui/core/Button"
import { useStyles } from "./Styles"
import Select from "react-select"

const CryptoList = () => {
  const globalStore = useSelector(state => state.globalStore)
  const dispatch = useDispatch()
  const classes = useStyles()
  const webSocket = useRef(null)

  useEffect(() => {
    dispatch(loadCurrencyList())
  }, [dispatch])

  const closeSnackbar = () => dispatch(handleSnackBarStatus(false))

  const handleInputChange = inputValue => {
    if (inputValue) {
      dispatch(handleUserCurrencySelection(inputValue.value))
    }
  }

  const ticker = globalStore.user_selected_currency.toLowerCase()

  return (
    <div className={classes.container}>
      {globalStore.loading ? (
        <div className={classes.spinner}>
          <LoadingSpinner />
        </div>
      ) : (
        <div className={classes.table}>
          <div className={classes.inputandButtonContainer}>
            <Select
              onChange={handleInputChange}
              className={classes.selectionBox}
              isClearable
              options={globalStore.currencyList}
              getOptionLabel={option => option.label}
              getOptionValue={option => option["value"]}
            />
            <Button
              onClick={() => {
                dispatch(getSelectedCurrency(ticker))
                webSocket.current = globalStore.current_websocket_connection
              }}
              className={classes.button}
              variant="contained"
              size="large"
              color="primary"
              disabled={globalStore.user_selected_currency === ""}
            >
              <Typography variant="h3">Proceed</Typography>
            </Button>
          </div>
        </div>
      )}
      <GlobalSnackbar
        open={
          globalStore.snackbar ||
          typeof globalStore.snackbar === "object" ||
          typeof globalStore.snackbar === "string" ||
          globalStore.snackbar instanceof String
        }
        variant="error"
        message={globalStore.error_while_fetching_initial_currency_list}
        onClose={closeSnackbar}
      />
    </div>
  )
}

CryptoList.propTypes = {
  onClose: PropTypes.func,
}

export default CryptoList
