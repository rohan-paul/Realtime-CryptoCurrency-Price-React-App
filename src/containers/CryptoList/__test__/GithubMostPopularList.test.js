import React from "react"
import renderer from "react-test-renderer"
import { MuiThemeProvider } from "@material-ui/core/styles"
import globalTheme from "../../../globalTheme"
import PropTypes from "prop-types"
import { mount, shallow } from "enzyme"
import expect from "expect"
import { createShallow, createMount } from "@material-ui/core/test-utils"
import configureMockStore from "redux-mock-store"
import thunk from "redux-thunk"
import { Provider } from "react-redux"
import CryptoList from "../CryptoList"
import EachUserListItem from "../EachUserListItem"
import {
  handleCityToSearchChange,
  handleSnackBarStatus,
} from "../../../actions/getUserActions"
import Button from "@material-ui/core/Button"
import { useDispatch } from "react-redux"
import * as Redux from "react-redux"
import { Action } from "redux"
import Autosuggest from "react-autosuggest"
import TablePagination from "@material-ui/core/TablePagination"

const middlewares = [thunk] // add your middlewares like `redux-thunk`
const mockStore = configureMockStore(middlewares)

describe("CryptoList Component", () => {
  let shallow: typeof enzyme.shallow
  let useSelectorSpy: jest.SpyInstance
  let useDispatchSpy: jest.SpyInstance

  let store
  let component
  let wrapperComp

  beforeEach(() => {
    shallow = createShallow()
    useSelectorSpy = jest.spyOn(Redux, "useSelector")
    useDispatchSpy = jest.spyOn(Redux, "useDispatch")
    useDispatchSpy.mockImplementation(() => (cb: Action) => cb)

    /* Everything you pass into mockStore will be your Redux store's initial state. So make sure you provide everything that's needed by your connected React component to render without any problems. */
    store = mockStore({
      globalStore: {
        loading: false,
        error_while_fetching_initial_currency_list: false,
        error_while_fetching_initial_data: false,
        city_to_search: "",
        snackbar: false,
        topTenUsersInCity: [],
        totalNoOfUsersFromAPI: 0,
      },
    })

    wrapperComp = mount(
      <Provider store={store}>
        <MuiThemeProvider theme={globalTheme}>
          <CryptoList />
        </MuiThemeProvider>
      </Provider>,
    )

    store.dispatch = jest.fn()

    component = renderer.create(
      <Provider store={store}>
        <MuiThemeProvider theme={globalTheme}>
          <CryptoList />
        </MuiThemeProvider>
      </Provider>,
    )
  })

  it("should render with given state from Redux store", () => {
    expect(component.toJSON()).toMatchSnapshot()
  })

  it("Bare minimum Shallow rendering for test should work", () => {
    useSelectorSpy.mockReturnValue({
      loading: false,
      error_while_fetching_initial_currency_list: false,
      error_while_fetching_initial_data: false,
      city_to_search: "",
      snackbar: false,
      topTenUsersInCity: ["a", "b"],
      totalNoOfUsersFromAPI: 0,
    })
    // const wrapperComp = shallow(<CryptoList />)

    const wrapperComp1 = shallow(
      <Provider store={store}>
        <MuiThemeProvider theme={globalTheme}>
          <CryptoList />
        </MuiThemeProvider>
      </Provider>,
    )
    expect(wrapperComp1.is(CryptoList))
  })

  it("should render EachUserListItem Component correct number of times", () => {
    const wrapper = mount(
      <Provider store={store}>
        <MuiThemeProvider theme={globalTheme}>
          <CryptoList />
        </MuiThemeProvider>
      </Provider>,
    )
    // console.log("STORE IS ", store.getState())
    expect(wrapper.find("EachUserListItem").length).toEqual(1)
  })

  it("should render Autosuggest Component correct number of times", () => {
    expect(wrapperComp.find("Autosuggest").length).toEqual(1)
  })

  /*   it("should submit proper search term after Autosuggest has value and the Search button is clicked", () => {
    wrapperComp.find("Autosuggest").props().inputProps.value = "Bangalore"
    expect(wrapperComp.find("Autosuggest").props().inputProps.value).toEqual(
      "Bangalore",
    )
    wrapperComp.update()
    console.log(
      "AUTO SUGGEST VALUE ",
      wrapperComp.find("Autosuggest").props().inputProps.value,
    )
    wrapperComp
      .find(Button)
      .last()
      .simulate("click")

    wrapperComp.update()

    console.log("STORE IS ", store.getState())
  }) */

  it("should submit proper search term after Autosuggest has value and the Search button is clicked", () => {
    wrapperComp.find(Autosuggest).props().inputProps.value = "Bangalore"
    expect(wrapperComp.find(Autosuggest).props().inputProps.value).toEqual(
      "Bangalore",
    )
    // wrapperComp.update()
    // console.log("AUTO SUGGEST VALUE ", wrapperComp.find("Autosuggest").props())

    const store1 = mockStore({
      globalStore: {
        loading: false,
        error_while_fetching_initial_currency_list: false,
        error_while_fetching_initial_data: false,
        city_to_search: `${
          wrapperComp.find(Autosuggest).props().inputProps.value
        }`,
        snackbar: false,
        topTenUsersInCity: [],
        totalNoOfUsersFromAPI: 0,
      },
    })

    const wrapperComp1 = mount(
      <Provider store={store1}>
        <MuiThemeProvider theme={globalTheme}>
          <CryptoList />
        </MuiThemeProvider>
      </Provider>,
    )

    wrapperComp1
      .find(Button)
      .last()
      .simulate("click")

    wrapperComp1.update()

    // console.log("STORE IS ", store1.getState())
    // console.log("NEW GITHUB COMP IS ", wrapperComp.props().store.getState())
    // console.log(
    //   "MODIFIED BUTTON IS ",
    //   wrapperComp1
    //     .find(Button)
    //     .last()
    //     .props(),
    // )
  })

  //************* */

  it("should dispatch an action on button click", () => {
    renderer.act(() => {
      wrapperComp
        .find(Button)
        .props()
        .onClick()
    })
    console.log("NEW GITHUB COMP IS ", wrapperComp.props().store.getState())
    console.log("USE DISPATCH ", useDispatchSpy)
    expect(useDispatchSpy).toHaveBeenCalledTimes(14)
    // expect(useDispatchSpy).toHaveBeenCalledWith(loadMostPopularUsers())
  })

  it("should render Button disabled props correctly ", () => {
    const button = wrapperComp.find(Button).last()
    button.simulate("click")
    expect(
      wrapperComp
        .find(Button)
        .last()
        .props().disabled,
    ).toEqual(true)
  })

  it("onClick props on Button should be defined", () => {
    const button = wrapperComp.find(Button).last()
    button.simulate("click")
    expect(
      wrapperComp
        .find(Button)
        .last()
        .props().onClick,
    ).toBeDefined()
  })

  // ***************
  // it("should dispatch the correct action on button click - 4", () => {
  //   const wrapperComp2 = mount(
  //     <Provider store={store}>
  //       <MuiThemeProvider theme={globalTheme}>
  //         <CryptoList />
  //         {/* <CryptoList loadAllData={getUsers} /> */}
  //       </MuiThemeProvider>
  //     </Provider>,
  //   )

  //   wrapperComp2.loadAllData = jest.fn()
  //   wrapperComp2.update()

  //   wrapperComp2
  //     .find(Button)
  //     .last()
  //     .simulate("click")
  //   wrapperComp2.update()
  //   // console.log("STORE IS ", store.globalStore.topTenUsersInCity)
  //   expect(wrapperComp2.loadAllData).toHaveBeenCalled()
  // })
})
