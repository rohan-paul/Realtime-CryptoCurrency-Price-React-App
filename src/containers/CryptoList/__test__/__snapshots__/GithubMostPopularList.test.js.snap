// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CryptoList Component should render with given state from Redux store 1`] = `
<div
  className="makeStyles-container-1"
>
  <div
    className="makeStyles-tableAndFabContainer-2"
  >
    <div
      className="makeStyles-table-3"
    >
      <div
        className="makeStyles-inputandButtonContainer-4"
      >
        <div
          aria-expanded={false}
          aria-haspopup="listbox"
          aria-owns="react-autowhatever-1"
          className="makeStyles-react_autosuggest__container-5"
          role="combobox"
        >
          <input
            aria-activedescendant={null}
            aria-autocomplete="list"
            aria-controls="react-autowhatever-1"
            autoComplete="off"
            className="makeStyles-react_autosuggest__input-6"
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            placeholder="Start typing your city name"
            type="text"
            value=""
          />
          <div
            className="makeStyles-react_autosuggest__suggestions_container-9"
            id="react-autowhatever-1"
            role="listbox"
          />
        </div>
        <button
          className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-containedSizeLarge MuiButton-sizeLarge Mui-disabled Mui-disabled"
          disabled={true}
          onBlur={[Function]}
          onClick={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={-1}
          type="button"
        >
          <span
            className="MuiButton-label"
          >
            <h3
              className="MuiTypography-root MuiTypography-h3"
            >
              Load City Data
            </h3>
          </span>
        </button>
      </div>
      <div
        style={
          Object {
            "marginTop": "20px",
          }
        }
      >
        <ul
          className="MuiList-root makeStyles-root-76 MuiList-dense MuiList-padding"
        />
      </div>
    </div>
    <div
      className="MuiTablePagination-root"
    >
      <div
        className="MuiToolbar-root MuiToolbar-regular MuiTablePagination-toolbar MuiToolbar-gutters"
      >
        <div
          className="MuiTablePagination-spacer"
        />
        <p
          className="MuiTypography-root MuiTablePagination-caption MuiTypography-body2 MuiTypography-colorInherit"
        >
          Rows per page:
        </p>
        <div
          className="MuiInputBase-root MuiTablePagination-input MuiTablePagination-selectRoot"
          onClick={[Function]}
        >
          <div
            aria-haspopup="listbox"
            aria-labelledby=" "
            className="MuiSelect-root MuiSelect-select MuiTablePagination-select MuiSelect-selectMenu MuiInputBase-input"
            onBlur={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onMouseDown={[Function]}
            role="button"
            tabIndex={0}
          >
            10
          </div>
          <input
            onAnimationStart={[Function]}
            type="hidden"
            value={10}
          />
          <svg
            aria-hidden="true"
            className="MuiSvgIcon-root MuiSelect-icon MuiTablePagination-selectIcon"
            focusable="false"
            role="presentation"
            viewBox="0 0 24 24"
          >
            <path
              d="M7 10l5 5 5-5z"
            />
          </svg>
        </div>
        <p
          className="MuiTypography-root MuiTablePagination-caption MuiTypography-body2 MuiTypography-colorInherit"
        >
          0-0 of 0
        </p>
        <div
          className="MuiTablePagination-actions"
        >
          <button
            aria-label="Previous page"
            className="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit Mui-disabled Mui-disabled"
            disabled={true}
            onBlur={[Function]}
            onClick={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            tabIndex={-1}
            title="Previous page"
            type="button"
          >
            <span
              className="MuiIconButton-label"
            >
              <svg
                aria-hidden="true"
                className="MuiSvgIcon-root"
                focusable="false"
                role="presentation"
                viewBox="0 0 24 24"
              >
                <path
                  d="M15.41 16.09l-4.58-4.59 4.58-4.59L14 5.5l-6 6 6 6z"
                />
              </svg>
            </span>
          </button>
          <button
            aria-label="Next page"
            className="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit Mui-disabled Mui-disabled"
            disabled={true}
            onBlur={[Function]}
            onClick={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            tabIndex={-1}
            title="Next page"
            type="button"
          >
            <span
              className="MuiIconButton-label"
            >
              <svg
                aria-hidden="true"
                className="MuiSvgIcon-root"
                focusable="false"
                role="presentation"
                viewBox="0 0 24 24"
              >
                <path
                  d="M8.59 16.34l4.58-4.59-4.58-4.59L10 5.75l6 6-6 6z"
                />
              </svg>
            </span>
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
